torch/mode_classes.py

```diff
@@ -0,0 +1,170 @@
+import functools
+
+# Implementation note: I had a choice about how much of mode stacks
+# to implement in Python versus in C++.  At time of writing, I did not care
+# too much about implementation efficiency; however, I do care about making it
+# hard for users to implement modes in the wrong way.  In the end, it turned
+# out to be possible to implement mode stacks entirely from userland, with the
+# C++ API providing only _get_torch_function_mode() and
+# _set_torch_function_mode(), so I opted to provide some unsafe C++ bindings and
+# have the bulk of the logic for managing the stack in Python, which helped
+# simplify the C++ API surface.  It would also have been valid to build in the
+# notion of mode stack directly into C++ but in this design it's substantially
+# more difficult to interact with TorchFunctionModeMeta.
+
+def _wrap_init(f):
+    undef = object()
+
+    @functools.wraps(f)
+    def wrapped(self, *args, inner=undef, **kwargs):
+        if inner is undef:
+            raise TypeError(
+                "missing inner keyword argument; instead of constructing a TorchModeFunction directly, "
+                "pass the constructor to push_python_mode"
+            )
+        self.inner = inner
+        return f(self, *args, **kwargs)
+    return wrapped
+
+def _wrap_torch_dispatch(f):
+    @functools.wraps(f)
+    def wrapped(self, *args, **kwargs):
+        with enable_python_mode(self.inner):
+            return f(self, *args, **kwargs)
+    return wrapped
+
+class ModeMeta(type):
+    def __new__(metacls, name, bases, dct):
+        if '__init__' in dct:
+            dct['__init__'] = _wrap_init(dct['__init__'])
+        return super().__new__(metacls, name, bases, dct)
+
+class PythonModeMeta(ModeMeta):
+    """
+    Metaclass for :class:`PythonMode`; it does two things:
+
+        * Adds an implicit ``inner`` kwarg to ``__init__``, to
+          allow the modes to be chained together to form a stack.
+
+        * Reenables the inner mode, so that by default PyTorch API calls
+          will compositionally proceed to the next mode on the stack.
+
+    The default behavior for the second bullet is important, as it is easy to
+    accidentally write ``_wrap_torch_dispatch`` implementations that are not
+    compositional, and the wrapping here makes the obvious code do the
+    right thing (aka, this is why there is a metaclass).
+    """
+    def __new__(metacls, name, bases, dct):
+        if '__torch_dispatch__' in dct:
+            dct['__torch_dispatch__'] = _wrap_torch_dispatch(dct['__torch_dispatch__'])
+        return super().__new__(metacls, name, bases, dct)
+
+
+class PythonMode(metaclass=PythonModeMeta):
```

Comment:
remove. These should be added by the next PR because their main use is for nested modes